File: src/main/java/dark/leech/text/plugin/js/loader/TextLoader.java
package dark.leech.text.plugin.js.loader;

import org.mozilla.javascript.NativeObject;

import dark.leech.text.action.Log;
import dark.leech.text.enities.PluginEntity;

/**
 * JavaScript chapter content loader for vBook plugins using Rhino. Executes chapGetter script and
 * returns chapter text as String.
 */
public class TextLoader extends AbstractLoader<String> {

    private TextLoader(PluginEntity plugin) {
        super(plugin);
    }

    public static TextLoader with(PluginEntity plugin) {
        return new TextLoader(plugin);
    }

    @Override
    protected String getScript() {
        return plugin.getChapGetter();
    }

    @Override
    protected LoaderType getLoaderType() {
        return LoaderType.TEXT;
    }

    @Override
    protected String processResult(Object result, String url) {
        // Check for Response wrapper (vBooks compatibility)
        if (!Response.isSuccess(result)) {
            String errorMsg = Response.getErrorMessage(result);
            Log.add("[TextLoader] Error response: " + errorMsg);
            return "";
        }

        Object data = Response.getData(result);

        // Handle different return types:
        // NativeObject with properties
        if (data instanceof NativeObject) {
            NativeObject obj = (NativeObject) data;

            Object body = obj.get("body", obj);
            String bodyStr = JSResponse.getString(body);
            if (bodyStr != null) {
                return bodyStr;
            }

            Object content = obj.get("content", obj);
            String contentStr = JSResponse.getString(content);
            if (contentStr != null) {
                return contentStr;
            }

            Object text = obj.get("text", obj);
            String textStr = JSResponse.getString(text);
            if (textStr != null) {
                return textStr;
            }
        }

        // Direct string result
        return JSResponse.getString(data);
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/js/loader/TextLoader.kt
package dev.haipham22.leechtext.plugin.js.loader

import dev.haipham22.leechtext.entities.PluginEntity
import dev.haipham22.leechtext.log.EngineLogger

/**
 * Loader nội dung chương (port từ loader/TextLoader.java; P5.2c: kết quả normalized).
 * Chạy script chapGetter, trả về text chương — hỗ trợ {body|content|text} object
 * hoặc string trực tiếp.
 */
class TextLoader private constructor(
    log: EngineLogger,
    plugin: PluginEntity,
) : AbstractLoader<String>(log, plugin) {
    companion object {
        fun with(
            plugin: PluginEntity,
            log: EngineLogger,
        ): TextLoader = TextLoader(log, plugin)
    }

    override fun getScript(): String? = plugin.chapGetter

    override fun getLoaderType(): LoaderType = LoaderType.TEXT

    override fun processResult(
        result: Any?,
        url: String?,
    ): String? {
        // Check Response wrapper (vBooks compatibility)
        if (!Response.isSuccess(result)) {
            val errorMsg = Response.getErrorMessage(result)
            log.add("[TextLoader] Error response: $errorMsg")
            return ""
        }

        val data = Response.getData(result)

        // Object với properties — lấy field đầu tiên có giá trị (body/content/text)
        val fromMap =
            (data as? Map<*, *>)
                ?.let { map -> listOf("body", "content", "text").firstNotNullOfOrNull { JSResponse.getString(map[it]) } }

        // String trực tiếp
        return fromMap ?: JSResponse.getString(data)
    }
}
